spatial_cell_gpu.cpp

```diff
@@ -0,0 +1,1170 @@
+    * to/from this cell.
+    *
+    * velocity_block_with_content_list needs to be up to date in local and remote cells.
+    * velocity_block_with_no_content_list needs to be up to date in local cells.
+    *
+    * update_velocity_block_with_content_lists() should have
+    * been called with the current distribution function values, and then the contetn list transferred.
+    *
+    * Removes all velocity blocks from this spatial cell which don't
+    * have content and don't have spatial or velocity neighbors with
+    * content.  Adds neighbors for all velocity blocks which do have
+    * content (including spatial neighbors).  All cells in
+    * spatial_neighbors are assumed to be neighbors of this cell.
+    *
+    * This function is thread-safe when called for different cells
+    * per thread. We need the block_has_content vector from
+    * neighbouring cells, but these are not written to here. We only
+    * modify local cell.*/
+
+   void SpatialCell::adjust_velocity_blocks(const uint popID, bool doDeleteEmptyBlocks) {
+      debug_population_check(popID);
+
+      const uint cpuThreadID = gpu_getThread();
+      const gpuStream_t stream = gpu_getStream();
+
+      phiprof::Timer adjustBlocksTimer {"Adjust velocity blocks"};
+      vmesh::VelocityMesh* host_vmesh    = populations[popID].vmesh;
+      vmesh::VelocityMesh* dev_vmesh    = populations[popID].dev_vmesh;
+      vmesh::GlobalID* _withContentData = velocity_block_with_content_list->data();
+
+      // Evaluate velocity halo for local content blocks
+      if (velocity_block_with_content_list_size>0) {
+         //phiprof::Timer blockHaloTimer {"Block halo kernel"};
+         const int addWidthV = getObjectWrapper().particleSpecies[popID].sparseBlockAddWidthV;
+         if (addWidthV!=1) {
+            std::cerr<<"Warning! "<<__FILE__<<":"<<__LINE__<<" Halo extent is not 1, unsupported size."<<std::endl;
+         }
+         // Halo of 1 in each direction adds up to 26 velocity neighbours.
+         // For NVIDIA/CUDA, we dan do 26 neighbours and 32 threads per warp in a single block.
+         // For AMD/HIP, we dan do 13 neighbours and 64 threads per warp in a single block, meaning two loops per cell.
+         // In either case, we launch blocks equal to velocity_block_with_content_list_size
+         update_velocity_halo_kernel<<<velocity_block_with_content_list_size, 26*32, 0, stream>>> (
+            dev_vmesh,
+            velocity_block_with_content_list_size,
+            _withContentData,
+            dev_velocity_block_with_content_map,
+            dev_velocity_block_with_no_content_map
+            );
+         CHK_ERR( gpuPeekAtLastError() );
+         //CHK_ERR( gpuStreamSynchronize(stream) );
+         //blockHaloTimer.stop();
+      }
+
+      /** Neighbour inclusion: GPU block adjusment now happens via batch calls.
+          Single-cell block adjustments will not want to account for neighbours, so this
+          code is no longer needed.
```

Comment:
Then drop it?

Reply:
Prefer to leave the code in for now, in case it becomes relevant at some point again (instead of then having to dig through old commits)